spec/factories/tag_factories.rb

```diff
@@ -12,13 +12,23 @@ def generate_oligo
   factory :v2_tag, class: Sequencescape::Api::V2::Tag do
     skip_create
 
+    sequence(:map_id) { |i| i }
     sequence(:oligo) { |_index| generate_oligo }
-    tag_group { create :v2_tag_group }
+    tag_group { create :v2_tag_group, v2_tags: [instance] }
   end
 
-  factory :v2_tag_group, class: Sequencescape::Api::V2::Tag do
+  factory :v2_tag_group, class: Sequencescape::Api::V2::TagGroup do
     skip_create
 
+    transient { v2_tags { [] } }
     sequence(:name) { |index| "TagGroup#{index}" }
+    tags { v2_tags.map { |t| { index: t.map_id, oligo: t.oligo } } }
```

Comment:
So `v2_tags` get passed into this factory as instances of `Sequencescape::Api::V2::Tag`, but this data structure seems to change them into a simple hash? Is that how it appears in real API results too?

Reply:
Yes.  I wouldn't have done it this way as it's not how V1 works, but the V2 endpoint was pre-existing with this and if I change it, who knows what else might break.

One of the biggest (and possibly least obvious) issues with APIs is that they're even harder to change than other aspects of systems.  You're setting up a contract with people outside your team's scope for how they can interface with your application.  Making decisions about API implementations should never be taken lightly.  Deploying a bad decision has to be maintained until the next version of the API.